### PR TITLE
Don’t enforce export/declare overload modifier consistency across module augmentations

### LINKED ISSUE #58756: Can't augment function overload is function is exported via `export {}`
### 🔎 Search Terms

TS2383, function overload, module augmentation

### 🕗 Version & Regression Information

- This is a crash
- Prior to 3.2.4 you'd get "TS2484: Export declaration conflicts with exported declaration of 'get'."

### ⏯ Playground Link

https://github.com/eps1lon/tsc-augment-overload

### 💻 Code

```ts
// library.d.ts
declare function get(): string;
declare function set(value: string): void;

export { get, set };

// app.d.ts
import "library";

declare module "library" {
  function get(): string | null;
}

// index.ts
import { get, set } from "library";

// module augmentation did work
const value = get();
//    ^? string | null
// @ts-expect-error Argument of type 'string | null' is not assignable to parameter of type 'string'.
set(value);
```

### 🙁 Actual behavior

```bash
$ yarn tsc --skipLibCheck false
app.d.ts:4:12 - error TS2383: Overload signatures must all be exported or non-exported.

4   function get(): string | null;
             ~~~


Found 1 error in app.d.ts:4
```

### 🙂 Expected behavior

No error since the module augmentation seems to have worked.

### Additional information about the issue

As a workaround, we can switch to 
```diff
-function get(): string
+export function get(): string
-export {get}
```
though this seems like TypeScript is enforcing a certain code style. Some teams prefer writing a single `export {}` statement.

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #58756

These rules are only there because it _looks_ confusing to have overload declarations like

```ts
// what does this mean??
declare function foo(): any;
function foo(s: number): string;
export declare function foo(s: string): number;
```

But if the overloads came from a module augmentation, it's reasonable for them to have different modifiers (see code comment and tests for the differences that are allowed).

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -40731,20 +40731,36 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
-                    else if (deviation & ModifierFlags.Ambient) {
-                        error(getNameOfDeclaration(o), Diagnostics.Overload_signatures_must_all_be_ambient_or_non_ambient);
-                    }
-                    else if (deviation & (ModifierFlags.Private | ModifierFlags.Protected)) {
-                        error(getNameOfDeclaration(o) || o, Diagnostics.Overload_signatures_must_all_be_public_private_or_protected);
-                    }
-                    else if (deviation & ModifierFlags.Abstract) {
-                        error(getNameOfDeclaration(o), Diagnostics.Overload_signatures_must_all_be_abstract_or_non_abstract);
+                group(overloads, o => getSourceFileOfNode(o).fileName).forEach(overloadsInFile => {
+                    const canonicalFlags = getEffectiveDeclarationFlags(getCanonicalOverload(overloads, implementation), flagsToCheck);
```
Comment (gabritto) [inline src/compiler/checker.ts:40735]: This computation can go outside the `forEach` since it doesn't change, right?

**Comment by andrewbranch** [comment]
@typescript-bot test it

**Comment by gabritto** [comment]
I feel like I don't have a good understanding of the old behavior to be able to assess the new one. Why was the old behavior in place? What are the bad scenarios we're trying to avoid, exactly?

**Comment by andrewbranch** [comment]
Stating my understanding, but I might have some of the details wrong. AFAIK, overloads are additional declarations of the same method/function symbol. Since there’s ultimately only one merged symbol, all these modifiers necessarily apply uniformly to the symbol. We don’t really have mechanisms in place to reason about what it would mean for some signatures to be exported while others aren’t. And the presence or absence of `declare` on an overload signature doesn’t really have any meaning at all, since overloads aren’t a runtime/emitting construct anyway. So I think the rules are in place because the mixing of modifiers like

```ts
export function foo(x: T1): T2;
export function foo(x: T3): T4;
       function foo(x: T5): T6;
export function foo(x: any): any {
  // ...
}
```

_looks_ like it might make the third overload file-local, but there’s actually nothing in place to make that happen, so we want to enforce consistency in the modifiers.

**Review by gabritto — APPROVED** [review]

**Pull request merged**